design/kanister-prometheus-integration.md

```diff
@@ -0,0 +1,290 @@
+
+## Motivation
+
+Kanister controller already has a registered metrics endpoint `/metrics`. 
+There are no metrics exported other than the default Prometheus metrics 
+that the default handler provides. Adding metrics to track the 
+ActionSets and Blueprints workflow will help improve the overall observability. 
+
+To achieve this, we need to build a framework for exporting metrics from the Kanister controller, and to 
+start with, export some metrics to Prometheus. 
+
+This framework simplifies the common need for Prometheus counters to
+publish 0 values at startup for all permutations of labels and label values.
+This ensures that Kanister controller restarts are recognized by Prometheus
+and that the PromQL rate() and increase() functions work properly across restarts.
+
+Some example metrics include:
+ActionSets succeeded
+ActionSets failed
+Action duration
+Phase duration, etc.
+
+
+## Scope
+
+1. Design a framework that allows us to export new
+   Kanister metrics to Prometheus easily. 
+2. Add a few fundamental metrics related to ActionSets and Blueprints to start with.
+
+
+## Detailed Design
+
+### Architecture
+
+![Alt text](Prometheus_Metrics_Design.png?raw=true "Prometheus Integration Design")
+
+#### Text description
+
+1. The initializer of the consumer package calls newMetrics, a helper method 
+   that talks to Kanister’s metrics package. The result is a new metrics struct
+   that owns all the Prometheus metrics.
+
+2. In order to initialize all the required Prometheus metrics, the new_metrics 
+   method calls the InitCounterVec, InitGaugeVec, InitHistogramVec in the 
+   metrics package. It passes the metric names and the specific label names 
+   and label values as BoundedLabels to the metrics package. Once it 
+   initializes all the Prometheus metrics successfully, it returns a struct that 
+   wraps all the metrics and that the consumer package can then use. 
+
+3. The metrics package internally initializes the Prometheus metrics and 
+   registers them Prometheus. If the registration fails because the specific 
+   metric with label header already exists, the metric will simply be returned 
+   to the caller. If the registration fails due to other reasons, then the 
+   metrics package will cause a panic, signaling programmer error. 
+   In case of the CounterVec, the InitCounterVec function generates all 
+   possible permutations of label values and initializes each counter 
+   within the CounterVec with a value of 0.
+
+4. Once the collector is created in the metrics package, it will be returned 
+   to the consumer package’s newMetrics helper method.
+
+5. Once the initialization of all Prometheus metrics are complete, a new 
+   metrics struct will be returned to the consumer’s package initializer. 
+
+6. The consumer package may find it useful to implement a helper method that constructs
+   a prometheus.Labels mapping to access a specific counter from a CounterVec 
+   and perform an increment operation.
```

Comment:
```suggestion
4. Once the collector is created in the metrics package, it will be returned to
   the consumer package’s `newMetrics` helper method.

5. Once the initialization of all Prometheus metrics are complete, a new
   metrics struct will be returned to the consumer package's initializer.

6. The consumer package may find it useful to implement a helper method that
   constructs a `prometheus.Labels` mapping to access a specific counter from a
   `CounterVec` and perform an increment operation.
```